Scratch Getting-started-guide

erictune opened the pull request.

erictune committed:
Scratch Getting-started-guide

erictune commented:
This is a bit rough around the edges, but it is ready for feedback from a number of people.
 - @mikedanese please take a look since you have tried setting up a cluster from scratch before.
 - @brendandburns  if you have time, feedback welcome.

erictune commented:
Fixes  #7871

erictune commented:
@ghodss

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,658 @@
+- Other nodes can have any reasonable amount of memory and any number of cores.  They need not
+  have identical configurations.
+
+### Selecting a Network
```
@thockin please review this section on networking for:
  - blatantly false statements
  - major missing information

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,658 @@
+
+If following the HTTPS approach, you will need to prepare certs and credentials.
+
+##### Preparing Certs
```
@roberthbailey please review this section on certs for:
  - blatantly false statements
  - major missing information

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,658 @@
+Started Guide and then copy the init.d scripts or systemd unit files from that
+cluster, and then modify them for use on your custom cluster.
+
+#### Configuring docker
```
@dchen1107  please review this section on kubelet for:
  - blatantly false statements
  - major missing information

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,658 @@
+    - `--api_servers=http://$MASTER_IP`
+    - `--kubeconfig=/var/lib/kube-proxy/kubeconfig`
+
+#### Node networking setup
```
@thockin  please review this section on networking for:
  - blatantly false statements
  - major missing information

erictune commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,658 @@
+  - make any modifications needed
+  - start the pod by putting it into the kubelet manifest directory
+
+#### Apiserver
```
@lavalamp @ArtfulCoder  please review this section on running master components for:
 - blatantly false statements
 - major missing information

lavalamp commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - `--cloud_provider=`
+  - `--cloud_config=` if cloud provider requires a config file (gce, aws). If so, need to put config file into apiserver docker image or mount through hostDir.
+  - `--address=${MASTER_IP}`.  
+   - or `--address=127.0.0.1` if you want to run a proxy on the master node.
```
Use --bind-address in addition to --address for this.

lavalamp commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - `--address=${MASTER_IP}`.  
+   - or `--address=127.0.0.1` if you want to run a proxy on the master node.
+  - `--cluster_name=$CLUSTER_NAME`
+  - `--service-cluster-ip-range=$SERVICE_CLUSTER_IP_RANGE`
```
Give an example range (10.0.1.0/24)?

erictune replied:
Added that example at line 83.

ddysher commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    - `--api_servers=https://$MASTER_IP`
+  - Otherwise, if taking the firewall-based security approach
+    - `--api_servers=http://$MASTER_IP`
+    - `--kubeconfig=/var/lib/kube-proxy/kubeconfig`
```
Why do we need kubeconfig for http instead of https?

ddysher commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    - `--api_servers=https://$MASTER_IP`
+  - Otherwise, if taking the firewall-based security approach
+    - `--api_servers=http://$MASTER_IP`
+    - `--kubeconfig=/var/lib/kube-proxy/kubeconfig`
```
Same question as to kube-proxy. 

One more thing:  looking at https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/saltbase/salt/kubelet/default
Looks like kubeconfig is not set for kubelet?

erictune replied:
Fixed.
You are right that the flag is not set in the default file.  The kubelet is using the default value of the flag.  I thought it was clearer to set it explicitly in this guide, so that the user understands the file needs to be present and so user can change path to file.

ddysher commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    serving its own API over HTTPS.
+
+Unless you plan to have a real CA generate your certs, you will need to generate a root cert and use that to sign the master, kubelet, and kubectl certs.
+    - see function `create-certs` in `cluster/gce/util.sh`
```
This doesn't format properly,  one less space before -?

erictune replied:
fixed.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+### Prerequisites
+  1. You should be familiar with using Kubernetes already.  We suggest you set
+    up a temporary cluster by following one of the other Getting Started Guides.
+    This will help you become familiar with the tools and API first.
```
We should point to which tools are being referred to here and also the API.

erictune replied:
updated.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    up a temporary cluster by following one of the other Getting Started Guides.
+    This will help you become familiar with the tools and API first.
+  1. You should have `kubectl` installed on your desktop.  This will happen as a side
+    effect of completing one of the other Getting Started Guides.
```
If we have them, we should point to the instructions on how to install kubectl.

erictune replied:
We do not have instructions for installing kubectl separate from the getting-started-guides.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+- Many Getting-started-guides make a distinction between the master node and regular nodes.  This
+  is not strictly necessary.
+- Nodes will need to run some version of Linux with the x86_64 architecture.  It may be possible
+  to run on other OSes and Architectures, but this guide does not try to assist with that.
```
I think "Machines with a version of x86_64 Linux" should also be listed under "Prerequisites". Can also make a similar statement to clarify that other OSes/Architectures are supported but not covered in this document.

erictune replied:
For node software, no other OSes or Architectures are supported.

For client software, we do support OSX and windows.  But that's not relevant in this section.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - Does not require "Routes" portion of Cloud Provider module.
+- Configure network to route Pod IPs
+  - Harder to setup from scratch.
+  - Implemented for GCE and AWS.
```
This bullet doesnt seem parallel with the other items. Do you mean that this option should be used if you want to eventually create clusters with nodes in GCE and AWS?

Also, we should define the first instance of GCE. Please change to "Google Compute Engine (GCE)".

erictune replied:
Reworded and spelled out GCE.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+This guide is also useful for those wanting to understand at a high level some of the
+steps that existing cluster setup scripts are making.
+
+### Prerequisites
```
Gathering from the sections below, we should also clarify and list the other items needed. For example, something like the following (please revise for clarity/accuracy):

3. Depending on your environment and requirements, you need to design and implement:
     - Networking model
     - Security model
4. For this getting started guide, the following hardware and software are required (these requirements are later discussed in detail):
     - A machine running x86_64 Linux for each node in the cluster.
     - *Mention whats needed or the assumptions made WRT "a working physical network"*
     - Download, install, and configure the following binaries:
          - Kubernetes (includes etcd, Docker, and rkt)

erictune replied:
I don't want to call those prerequisites, because that implies that people should go prepare those things before working further.  But, I don't want to them to do that.  I want them to read further to understand the different ways in which they could do those things before doing them.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    This will help you become familiar with the tools and API first.
+  1. You should have `kubectl` installed on your desktop.  This will happen as a side
+    effect of completing one of the other Getting Started Guides.
+
```
Can we add a section title here: "Designing and Preparing"

And then add a TOC after the new title to define/link to each section:

 - Cloud Providers
 - Nodes
 - Network Models
 - Software Binaries
 - Naming Clusters
 - Security Models

erictune replied:
done.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+guide assume kubeconfigs in `/var/lib/kube-proxy/kubeconfig` and
+`/var/lib/kubeletproxy/kubeconfig`.
+
+#### Selecting Binaries
```
Can we move this section before the "Selecting a Cluster Name" section?

-> The "Selecting a Security Model" section refers to k8s files that they wont have unless we ask them to figure out what binaries to use earlier.

erictune replied:
done

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - kube-controller-manager
+  - kube-scheduler
+
+You can use a binary release (recommended) or build your own following the instructions in the
```
Please insert "Kubernetes" -> "...a Kuberenetes binary..."

erictune replied:
done

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+You will need binaries for:
+  - etcd
+  - docker or rkt
+  - kubelet
```
Can we add buckets to the list (one for containers and another for k8s):

 - etcd
 - A container image, either:
     - Docker
     - rkt
 - Kuberenetes:
    - kubelet
    - ...

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+You can use a binary release (recommended) or build your own following the instructions in the
+[Developer Documentation]( ../devel/README.md).  Only using a binary release is covered in this guide.
+
```
Can we add a section header here and also mention that all the binaries you list above are bundled with the k8s binaries?

#####Downloading and Extracting Kubernetes Binaries

The Kubernetes binary release includes all the Kubernetes binaries as well as the supported releases of etcd, Docker, and rkt.  (*is that right, both Docker and rkt are bundled?*)

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+The kube-apiserver, kube-controller-manager, and kube-scheduler, you will run as pods,
+so there needs to be a docker image in whatever docker repository you are using.
+Options:
```
rkt is not mentioned here. Are the rkt images in GCR too? Otherwise we should add where and how to get those too?

yifan-gu replied:
I don't think there are ACIs on GCR yet. But I think we have built images for these components internally.
/cc @philips @kelseyhightower @bcwaldon

erictune replied:
I think rkt documentation is going to need to be expanded later.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+You may also run etcd in a pod, in which case you can:
+  - Use a GCR image, such as `gcr.io/google_containers/etcd:2.0.12`
+  - Build your own image.
```
Can we point to the instructions on how to "build your own image"?

erictune replied:
Updated.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - tokens are just long alphanumeric strings, e.g. 32 chars.  See
+    - `TOKEN=$(dd if=/dev/urandom bs=128 count=1 2>/dev/null | base64 | tr -d "=+/" | dd bs=32 count=1 2>/dev/null)`
+
+Any tokens and.or passwords need to stored in a file for the apiserver
```
Revise to:
Your tokens and passwords need to be stored....

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+You will also need to do assorted other configuration on top of a
+base OS install.
+
+One possible starting point is to setup a cluster using an existing Getting
```
Prefix with "Tip": 
Tip: One possible......

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
````diff
@@ -0,0 +1,664 @@
+current-context: service-account-context
+```
+Put the kubeconfig(s) on every node.  The examples later in this
+guide assume kubeconfigs in `/var/lib/kube-proxy/kubeconfig` and
````
Revise to:
...assume that there are kubeconfig files in ....

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+base OS install.
+
+One possible starting point is to setup a cluster using an existing Getting
+Started Guide and then copy the init.d scripts or systemd unit files from that
```
Revise to:
...Started Guide. After getting a cluster running, you can then copy the......

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - `SCHEDULER_IMAGE=gcr.io/google_containers/kube-scheduler:$TAG`
+  - `CNTRLMNGR_IMAGE=gcr.io/google_containers/kube-controller-manager:$TAG`
+
+### Configuring Base Node Software
```
Revise to:
Configuring Base Software on Nodes

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+installtion, by following examples given in the Docker documentation.
+
+Remember to eliminate any network configuration created as part of running docker previously
+with different options:
```
Is this run once (after its "working correctly") or run after each set of Docker options that you try?

 - For run once at the end (after you get it working), revise sentence to:
"To eliminate any network configuration created from previously running docker with different options, run the following commands:"
 - For run after each set of Docker options:
     - We should move this sentence up above the sentence at line 281 ("Ensure Docker is working....")
     - Revise to:
"Remember that for each set of Docker options you try, you have to first clean up the network configuration before trying different options. To remove the previous network configuration, run the following commands:"

erictune replied:
One time. Clarified.  Moved higher up. in section.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+   - `DOCKER_NOFILE=1000000`
+
+Ensure docker is working correctly on your system before proceeding with the rest of the
+installtion, by following examples given in the Docker documentation.
```
typo: "installation"

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+- Open any firewalls which would hinder communication between the nodes.
+- Open any firewalls to allow access to the apiserver ports 80 and/or 443.
+
+### Other preparations
```
I think we can remove this (once the "Designing and Preparing" section header is added and section TOC inserted)

erictune replied:
Did those thing but this content still needs some home.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+further in the [networking documentation](../networking.md). The recommended way to 
+do this is to let kubelet do it by passing `--configure-cbr0=true` to kubelet.
+Alternatively, the bridge can be created manually, like this:
+- Create a bridge
```
Please make ordered list (not bullets)

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+into a [kubeconfig file](../kubeconfig-file.md).
+
+The kubeconfig file for the administrator can be created as follows:
+ - If you have already used Kubernetes with a non-custom cluster, you will already have a `$HOME/.kube/config` file.
```
Insert after "non-custom cluster": "(for example, used a Getting Started Guide)"
... a non-custom cluster (for example, used a Getting Started Guide),...

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+will ensure that responses destined to the node still reach
+the pod.
+
+#### Other configuration
```
Include pointers to the "how to" info for the bullets that dont already have links

erictune replied:
These are things that people undertaking an installation from scratch are expected to know how to do.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+#### Using Configuration Management
+The previous steps all involved "conventional" system administration techniques for setting up
+machines.  You may want to use a Configuration Management system to automate the node configuration
+process.  There are examples of Saltstack, Ansible, Juju, and CoreOS Cloud Config in the
```
Can we point to our k8s saltstack config info? Is it accurate for "starting from scratch"?

If we can use it, please use the relative link:  
 ```
[Saltstack](../../docs/salt.md)
```

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+    pods`
+
+#### etcd
+You will need to run one or more instance of etcd.  
```
Typo: instances (plural)

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+ See [Availability](../availability.md) for more discussion on factors affecting cluster
+availability.
+
+To run an etcd instance:
```
Please make ordered list (not bullets)

erictune replied:
ok

mikedanese replied:
This section is not formatted properly in the markdown.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - start the pod by putting it into the kubelet manifest directory
+
+#### Apiserver
+To run the apiserver:
```
Please make ordered list (not bullets)

erictune replied:
ok

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+*TODO* document proxy-ssh setup.
+
+##### Apiserver pod template
```
Is there some guidance we can point to or provide here for this apiserver template and the following templates below (for scheduler and controller manager)? For example, which ones are important (which are not and can be ignored), how to set them, etc.

RichieEscarez replied:
Move this whole section about the template under the "write a pod spec for the apiserver using the provided template" (399), so these options show under the corresponding step in the process.

erictune replied:
I can't devote more time to improving this advice right now.  Remember that this guide is for people who have decided not to use one of the many automated setup processes we provide, so the expectation of the level of detail is not the same as a normal document.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+Optionally, you may want to mount `/var/log` as well and redirect output there.
+
+#### Starting Apiserver
+Place the completed pod template into the kubelet config dir
```
Can we make these steps an ordered list to highlight each step?

erictune replied:
See previous comment.

erictune replied:
1. it looks funny  to have
```
block quoted code
```
2. inside and ordered list
3. do you agree?
4. and the numbers are wrong.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+various Getting Started Guides.
+
+### Bootstrapping the Cluster
+
```
Can we insert a TOC (with links) of this section here?

 - Etcd
 - Apiserver
 - Scheduler
 - Controller Manager
 - Resources and Addons

erictune replied:
I put a TOC at the top.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+  - write a pod spec for the apiserver using the provided template
+  - start the pod by putting it into the kubelet manifest directory
+
+Here are some apiserver flags you may need to set:
```
Move this whole section about flags under the "select the correct flags for your cluster" (398), so these options show under the corresponding step in the process.

erictune replied:
I'm not going to rewrite this section in this PR.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+To run the apiserver:
+  - select the correct flags for your cluster
+  - write a pod spec for the apiserver using the provided template
+  - start the pod by putting it into the kubelet manifest directory
```
Hyperlink this step to the instructions below at line 494 (starting the apiserver....).

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+You should soon be able to see all your nodes by running the `kubect get nodes` command.
+Otherwise, you will need to manually create node objects.
+
+#### Scheduler
```
The formatting here should be similar to the changes ive requested for under the apiserver section, where:

 - There is an ordered list of the steps.
 - Under each step is the corresponding configuration info. (one step for scheduler template since there are no flags mentioned here)
 - The last step is a hyperlink pointing to the "Starting the apiserver, scheduler, controller manager" section.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+Optionally, you may want to mount `/var/log` as well and redirect output there.
+
+#### Starting Apiserver
```
Can we generalize the steps and clarify the title here (ie steps also used to start scheduler). How about controller manager?

####Starting the Apiserver, Scheduler, or Controller Manager
Use these steps to start the apiserver, scheduler, and controller manager, after you create the configuration templates for each (ran in the order that's specified in this guide).

erictune replied:
done in follow up PR.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+Start as described for apiserver.
+
+#### Controller Manager
```
Please reformat similar to whats explained in comment at line 522

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+#### Create Miscelaneous Resources and Addons
+
+##### Miscelaneous Resources
```
Can we also link to the topic and examples we have about resourcequotas and limitranges (if they are applicable to "starting from scratch")?

RichieEscarez commented:
@erictune, sorry so many comments (mostly formatting suggestions and requests for adding links to related info), but im done reviewing.

yifan-gu commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+This section discusses how to configure machines to be kuberetes nodes. 
+
+You should run three daemons on every node:
+  - docker or rkt
```
rkt isn't a daemon, but we will need [rkt metadata service](https://github.com/appc/spec/blob/dcbaa9dbc4036c501dde9742378411d6b5a5a36c/spec/ace.md#app-container-metadata-service)

erictune replied:
acknowledged.  Once this is in, feel free to improve the explanation around using rkt.

erictune commented:
@RichieEscarez 
- I write "docker" when I am talking about the docker daemon or client binary or process, and *Docker* when I am talking about the overall product.  I've fixed a few places where I was inconsistent about that.
- Fixed all the capitalization mistakes I could find in a quick pass.
- I agree the lists could be better worded.  I'm going to leave further improvements to the community, though, and focus on further content generation.

ArtfulCoder commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,664 @@
+
+### Selecting a Network
+Kubernetes has a distinctive [networking model](../networking.md) which uses
+"virtual" IPs for pods and services and a different virtual ethernet device for each pod.
```
kubernetes requires that containers across different nodes be "ping"-able from each other using their ips. If that is not the case, an overlay network, like Flannel has to be used. 
Kubernetes creates virtual IPs for services, not for pods.

erictune replied:
Reworded.

erictune commented:
Addressed all review comments in one way or another.  Squashed. 
Looking for an LGTM at this point.

ArtfulCoder commented:
@erictune Did you get a chance to look at my comment "kubernetes requires that containers across different nodes be "ping"-able from each other using their ips. If that is not the case, an overlay network, like Flannel has to be used.  Kubernetes does not create virtual IPs for pods."

ghodss commented:
Some of our team looked over this and thought it looked pretty good, and definitely would have accelerated some of our packaging efforts, so that's awesome. Thanks for working on this.

One suggestion would be to have a "Upgrade notes" section which would include any changed flags, file locations, settings, etc. per released version. In theory you could git diff this document but I think an explicit section would help a lot when we go from 0.18 to 0.19, 0.19 to 1.0, etc.

erictune commented:
@ArtfulCoder I did change the doc in response to your comment.

@ghodss I don't have the time to do this retroactively, but I'll try to remember to update this doc in the future.

erictune commented:
@mikedanese can I get an LGTM.

erictune commented:
shippable is a timeout.  Should be safe to merge.

ghodss commented:
@erictune SGTM

mikedanese commented:
I'm going to try to read this tonight. Lot's of stuff going on today.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+### Cloud Provider
+Kubernetes has the concept of a Cloud Provider, which is a module which provides
+an interface for managing TCP Load Balancers, Nodes (Intances) and Newtworking Routes.
```
"Intances" and "Newtworking"

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+- Nodes will need to run some version of Linux with the x86_64 architecture.  It may be possible
+  to run on other OSes and Architectures, but this guide does not try to assist with that.
+- Apiserver and etcd together are fine on a machine with 1 core and 1GB RAM for clusters with 10s of nodes.
+  Larger or more active clusters may benefit from a more cores.
```
"a more cores"

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+Kubernetes has a distinctive [networking model](../networking.md).
+
+Kubernetes allocates an IP address to each pod and creates a virtual ethernet device for each
+Pod.  When creating a cluster, you need to allocate a block of IPs that Kubernetes to be used
```
"that Kubernetes to be used"

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  - Traffic is encapsulated, so per-pod IPs are routable.
+  - Examples:
+    - [Flannel](https://github.com/coreos/flannel) 
+    - [Weave](http://weave.works/)
```
might want to mention ovs, which is used by vagrant and openshift guys I believe. It is by far the most mature option compared to flannel and weave.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+Also, you need to pick a static IP for master node.
+- Call this `MASTER_IP`.
+- Open any firewalls to allow access to the apiserver ports 80 and/or 443.
```
Don't forget to enable ipv4 forwarding! `net.ipv4.ip_forward = 0`

erictune replied:
Seems like docker does this for you now? https://docs.docker.com/articles/networking/ says the default value of `--ip-forward` is true and when true docker sets this sysctl?

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+Download the [latest binary release](
+https://github.com/GoogleCloudPlatform/kubernetes/releases/latest) and unzip it.
+Then locate `./kubernetes/server/kubernetes-server-linux-amd64.targz` and unzip *that*.
```
nit : s/targz/tar.gz

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+    `tar -C kube-apiserver -c . | docker import - kube-apiserver`
+  - *TODO*: test above command.
+
+For
```
I think you want "For etcd: "

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+    - KUBELET_CERT
+      - optional
+    - KUBELET_KEY
+      - optional
```
This blurb is not formatted correctly in markdown. I think you might need to unindent once. Also the last three envvars don't have ``.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  - tokens are just long alphanumeric strings, e.g. 32 chars.  See
+    - `TOKEN=$(dd if=/dev/urandom bs=128 count=1 2>/dev/null | base64 | tr -d "=+/" | dd bs=32 count=1 2>/dev/null)`
+
+You tokens and passwords need to stored in a file for the apiserver
```
s/You/Your

mikedanese replied:
"need to stored"

mikedanese commented on docs/getting-started-guides/scratch.md:
````diff
@@ -0,0 +1,727 @@
+```
+Put the kubeconfig(s) on every node.  The examples later in this
+guide assume that there are kubeconfigs in `/var/lib/kube-proxy/kubeconfig` and
+`/var/lib/kubeletproxy/kubeconfig`.
````
kubeletproxy?

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+## Configuring and Installing Base Software on Nodes
+
+This section discusses how to configure machines to be kuberetes nodes. 
```
"kuberetes"

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+      cluster.
+    - if you are using an overlay network, consult those instructions.  
+  - `--bip=` and `--mtu=`
+    - may be required when using Flannel; see Flannel instructions.
```
mtu may be required when using flannel because of udp encapsulation but i believe --bip should always be the cidr of the PodNet for that specific node, otherwise the bridge will come up with the default docker cidr.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  - `--bridge cbr0`
+  - `--iptables=false`
+  - `--ip-masq=false`
+    - if have setup PodIPs to be routable, then you want this false, otherwise, docker will
```
"if have setup PodIPs to be". also --iptables=false then --ip-masq is ignored whatever it is set to.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+You may want to increase the number of open files for docker:
+   - `DOCKER_NOFILE=1000000`
+
```
Where do i put all these configs^^^? mention /etc/default/docker?

thockin replied:
This is still open afaict - something like "this will vary based on your node's OS, for example GCE's debian distro use /etc/default/docker"

mikedanese commented on docs/getting-started-guides/scratch.md:
````diff
@@ -0,0 +1,727 @@
+If you have turned off docker ip masquerading to allow pods to talk to each
+other, then you may need to do masquerading just for destination IPs outside
+the cluster network.  For example:
+```iptables -w -t nat -A POSTROUTING -o eth0 -j MASQUERADE \! -d ${CONTAINER_SUBNET}```
````
found the postrouting rule!

thockin replied:
CONTAINER_SUBNET is undefined.  Need to explain it as something like "the range of addresses which can be accessed by pod IPs" and maybe further that some environments will not egress traffic from IPs that are not natively known (e.g. a VM IP is known by GCE's fabric but pod IPs are not).

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+    - `--api_servers=http://$MASTER_IP`
+
+### Networking 
+Each node needs to be allocated its own CIDR range for pod networking.
```
"allocated its own"

erictune replied:
not sure what you are asking for here.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  1. Set `--configure-cbr0=true` option in kubelet init script and restart kubelet service.  Kubelet will configure cbr0 automatically.
+     It will wait to do this until the node controller has set Node.Spec.PodCIDR.  Since you have not setup apiserver and node controller
+     yet, the bridge will not be setup immediately.
+- Alternate, manual approach:
```
You also need to configure docker to use this bridge in this approach with `--bridge=cbr0`

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  - their options are specified in a Pod spec (yaml or json) rather than an /etc/init.d file or
+    systemd unit.
+  - they are kept running by Kubernetes rather than by init.
+  - they run in containers, and so tools like `ps` and `top` are less useful than `docker ps` and `kubectl get
```
ps and top on the host will still show processes running in containers e.g. try running `ps ax | grep kube-apiserver` on a master

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+To run the apiserver:
+1. select the correct flags for your cluster
+1. write a pod spec for the apiserver using the provided template
+1. start the pod by putting it into the kubelet manifest directory
```
This section is not formatted properly in the markdown.

mikedanese commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+
+**TODO** talk about starting DNS.
+
+### Miscelaneous Resources
```
Miscellaneous is spelled (and in the index) wrong and this title sounds like a section where I can get more information beyond this guide.

mikedanese commented:
Most are small changes, these is pretty much LGTM as a first pass. It's a good start and we should get it in so people can start addressing the TODOs and fixing up/adding to the various pieces.

RichieEscarez commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+### Controller Manager
+To run the controller manager:
+  - select the correct flags for your cluster
+  - write a pod spec for the apiserver using the provided template
```
Please update to "controller manager":
The line above should say "...pod spec for the controller manager..." (not apiserver).
&
The line below should say "start the controller manager pod" (not scheduler)

erictune commented:
Fixed most of @mikedanese comments and @RichieEscarez comment.
I think this is ready to go in.

mikedanese commented:
LGTM assigining to @RichieEscarez

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+
+### Nodes
+- You can use virtual or physical machines.
+- The machines need to be on the same subnet.
```
Define subnet?  Are you meaning an IP-space?  Technically we're all on a /0 subnet, I guess.  I think this statement is not clear asis.

erictune replied:
Please suggest better language.

erictune replied:
never mind, removed line.

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+    - [Open vSwitch (OVS)](http://openvswitch.org/)
+  - Does not require "Routes" portion of Cloud Provider module.
+
+You need to select an address range for the virtual IPs. 
```
s/virtual/pod/ here - pod IPs are not virtual, even if the "devices" they are attached to are.

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+  - GCE: each project has own `10.0.0.0/8`.  Carve off a `/16` from that.  Room for several clusters in there.
+  - AWS: use one VPC for whole organization, carve off a chunk for each cluster.  Or use different VPC for different clusters.
+  - IPv6 not supported yet.
+- Allocate one CIDR for PodIPs for each node, or a large CIDR from which
```
Maybe mention that a /24 per node is common, but probably overkill.  If you are tight on IPs a /25 or /26 or even /27 is probably enough (126, 62, and 30 pods, respectively)

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+Kubernetes also allocates an IP to each [service](../services.md).  However, service IPs do not necessarily
+need to be routable.  The kube-proxy takes care of translating Service IPs to Pod IPs before traffic leaves
+the node.  You do need to Allocate a block of IPs for services.  Call this `SERVICE_CLUSTER_IP_RANGE`.
+ Example: `SERVICE_CLUSTER_IP_RANGE="10.0.1.0/24"`
```
Maybe example as /16 and note that the width of this CIDR determines how many Service objects can be active at once.  Also maybe note that it may be possible to expand this into adjacent space at a later time, it is not possible to move it to a different address range without killing all Services.

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+Also, you need to pick a static IP for master node.
+- Call this `MASTER_IP`.
+- Open any firewalls to allow access to the apiserver ports 80 and/or 443.
+- Enable ipv4 forwarding, `net.ipv4.ip_forward = 0`
```
you want `= 1`

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+
+The way you configure docker will depend in whether you have chosen the routable-vip or overlay-network approaches for your network.
+Some docker options will want to think about:
+  - `--bridge cbr0`
```
explain: assuming you created your own bridge for the per-node CIDR ranges (which is what GCE does).  You could also use the --bip flag

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+Some docker options will want to think about:
+  - `--bridge cbr0`
+  - `--iptables=false`
+    - so that kube-proxy can manage iptables instead of docker.
```
description is not right.  In fact we probably want to stop setting this flag soon.  Better described as

"don't manipulate iptables for host-ports (too coarse on older docker versions, may be fixed in newer versions)"

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+  - `--iptables=false`
+    - so that kube-proxy can manage iptables instead of docker.
+  - `--ip-masq=false`
+    - if you have setup PodIPs to be routable, then you want this false, otherwise, docker will
```
NB: Some environments (e.g. GCE) still need you to masquerade out-bound traffic when it leaves the cloud environment.  This is very environment specific.

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,726 @@
+
+Arguments to consider:
+  - If following the HTTPS security approach:
+    - `--api_servers=https://$MASTER_IP`
```
spec flags with dashes rather than underscores?  Our daemons should all support both, but dashes are the "right" way

erictune commented:
Fixed all @thockin comments.
No further reviews requested.
Can I please get LGTMs from @thockin @mikedanese

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+Kubernetes also allocates an IP to each [service](../services.md).  However, service IPs do not necessarily
+need to be routable.  The kube-proxy takes care of translating Service IPs to Pod IPs before traffic leaves
+the node.  You do need to Allocate a block of IPs for services.  Call this `SERVICE_CLUSTER_IP_RANGE`.
+For example, you could set `SERVICE_CLUSTER_IP_RANGE="10.1.0.0/16"`, allowing ~67840 distinct services to be active at once.
```
Where does 67840 come from?  Shouldn't it be 65534?

thockin commented on docs/getting-started-guides/scratch.md:
```diff
@@ -0,0 +1,727 @@
+  1. Set appropriate MTU
+  - `ip link set dev cbr0 mtu 1460`
+  1. Add the clusters network to the bridge (docker will go on other side of bridge).
+  - e.g. `ip addr add $CLUSTER_CIDR dev eth0`
```
CLUSTER_CIDR is undefined in this doc.  $NODE_X_POD_CIDR is closer, but not right - you need an IP in that CIDR.

e.g. if `$NODE_X_POD_CIDR=10.9.8.0/24`, you want `10.9.8.1/24`

thockin commented:
This is a great doc and I don't want to hold it up, so I'll LGTM now.

thockin commented:
I'm going to merge to get this in and rev on it.

RichieEscarez commented:
I missed seeing this was merged but ive opened #10767 for all the editorial comments in this thread that were not part of the merge

thockin commented:
Sorry, I merged so I could fix my own comments :)

erictune commented:
Linked to kubectl install insrtructions in follow up PR.
